datafusion/expr/src/logical_plan/invariants.rs

```diff
@@ -112,9 +112,11 @@ fn assert_valid_semantic_plan(plan: &LogicalPlan) -> Result<()> {
 /// Returns an error if the plan does not have the expected schema.
 /// Ignores metadata and nullability.
 pub fn assert_expected_schema(schema: &DFSchemaRef, plan: &LogicalPlan) -> Result<()> {
-    let equivalent = plan.schema().equivalent_names_and_types(schema);
+    let compatible = plan
+        .schema()
+        .check_arrow_schema_type_compatible(&schema.as_ref().into());
 
-    if !equivalent {
+    if compatible.is_err() {
```

Comment:
can you please also add the `compatiple` message here? 

Perhaps something like

```
            "Failed due to a difference in schemas: {e}, original schema: {:?}, new schema: {:?}",
```